snuba/migrations/migrations.py

```diff
@@ -0,0 +1,91 @@
+import logging
+
+from datetime import datetime
+from enum import Enum
+from importlib import import_module
+from typing import Optional, Sequence, Tuple
+
+from snuba.clusters.cluster import ClickhouseClientSettings, get_cluster, CLUSTERS
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations.operations import Operation
+
+
+logger = logging.getLogger("snuba.migrations")
+
+LOCAL_TABLE_NAME = "migrations_local"
+DIST_TABLE_NAME = "migrations_dist"
+
+# Currently only local migrations are supported
+TABLE_NAME = LOCAL_TABLE_NAME
+
+
+class App(Enum):
+    SYSTEM = "system"
+    SNUBA = "snuba"
+
+
+class Status(Enum):
+    NOT_STARTED = "not_started"
+    IN_PROGRESS = "in_progress"
+    COMPLETED = "completed"
+
+
+class Migration:
+    is_dangerous: bool
+    dependencies: Optional[Sequence[Tuple[str, str]]]
+
+    def forwards_local(self) -> Sequence[Operation]:
+        raise NotImplementedError
+
+    def backwards_local(self) -> Sequence[Operation]:
```

Comment:
What is the semantics of the backwards sequence of operations and which guarantees do users have to provide when writing a migration ? Specifically:
- Is it up to the user to guarantee that all Operations are reversible? Or does that apply to migrations as a whole and not individually to operations ?
- when a migration fails and only some of the operations are completed, is it the job of the backwards operations to bring us back to a stable state or is it the job of the forward operations to pick up from things went badly ? If it is the second, what would we use the backwards operations for ?
- Will it be possible to revert a completed migration? I don't think so because there could be user data.

Reply:
I think the primary purpose will be the second one you mentioned - to bring us back to the stable state if some operation within the migration failed. It could also be useful to make sure our databases are in the right state in development.
I think that our schema migrations should generally always have a backwards migration but the data migration parts could be no-op.

Reply:
> I think that our schema migrations should generally always have a backwards migration

I am not sure this is viable. If a migration deletes a column, how would you restore it? I think we should limit this scope unless there is a strong reason to revert entire migrations. For example only an "in progress" migration can be reverted via the backwards steps. Or is there a strong case to support reverting entire migrations?

Agree otherwise on putting the responsibility to bring us back to a stable state to the backwards steps. Here as well, the semantics should be clearly documented in the docstring to explain what a migration can do, cannnot do and what the responsibilities of the developer are.

Reply:
I don't know if it will be eventually valuable to expose the ability to go backward via the CLI or we should not expose it at all. It could be useful if say we realise something is wrong immediately after a migration and we just want to revert the changes we made. For many schema migrations we'll be able to go backwards correctly. But I don't know if this is a likely scenario, and it's never going to be a guaranteed way to get back the prior state since we might be missing data.

Reply:
When something works in most cases, but silently does not work or does not work properly in some crucial ones compromising correctness, I think it is better not to support it at all or figure out a way to make it evident what is supported and what is not.

Providing a way to revert entire migrations where it is possible and safe could be reasonable, but we would have to make it safe by introducing a flag in the migration that says whether the migration is reversible. And ensure we never try to revert an irreversible migration.